### ISSUE TITLE
Cubic reciprocity law

### ISSUE THREAD

**Issue by Bodigrim** [cartazio/arithmoi#155]
It may be fun to extend [Math.NumberTheory.Moduli.Jacobi](https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/Moduli/Jacobi.hs) to cover not only [quadratic reciprocity](https://en.wikipedia.org/wiki/Quadratic_reciprocity), but [cubic](https://en.wikipedia.org/wiki/Cubic_reciprocity) as well.

Quadratic reciprocity symbol or [Jacobi symbol](https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/Moduli/Jacobi.hs#L33) answers the question whether an equation `x^2 = n (mod m)` has any solutions. Similarly, cubic residue symbol answers whether `x^3 = n (mod m)` has solutions.

The values of Jacobi symbol are units of real integers (+ zero), while the values of cubic symbol are units of Eisenstein integers.

When this will be implemented, we can proceed to modular cubic roots (similar to [Math.NumberTheory.Moduli.Sqrt](https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/Moduli/Sqrt.hs)) and eventually solve cubic and quartic modular equations in [Math.NumberTheory.Moduli.Equations](https://github.com/cartazio/arithmoi/blob/master/Math/NumberTheory/Moduli/Equations.hs).

**Comment by hgsipiere** [comment]
I was looking into this issue out of curiosity and noticed that with cubic/quartic residue characters the values are roots of unity. Would it make sense to move roots of unity out of Dirichlet characters or to just reference the Dirichlet character module if implementing this?

Edit: Ignore this, can just use EisensteinIntegers module

**Comment by Bodigrim** [comment]
Moving them out is a good idea, I think.

**Comment by b-mehta** [comment]
I agree roots of unity probably belong somewhere other than in Dirichlet characters, but cubic/quartic symbols should have a more specific type than root of unity. It would be nice to include a conversion function `CubicResidueSymbol -> RootOfUnity` or something similar instead.

**Comment by hgsipiere** [comment]
I completely agree, representing them as Eisenstein Integers looks nice with stuff like CubicResidueSymbol ω (or anything else more specific). It was an oversight, I think they are seperate issues.

Edit: I think you can only do CubicResidueSymbol -> Maybe RootOfUnity because of the 0 case.

**Comment by Bodigrim** [comment]
`CubicSymbol` is also fine from my perspective. But representing cubic symbols directly as `EisensteinInteger` is probably painful because of the lack of exhausting pattern matching.

IMO we can iron out precise API details later, having more stuff at hand.
